app/src/main/java/com/example/drawingApp/MainActivity.kt

```diff
@@ -1,23 +1,114 @@
 
 class MainActivity : AppCompatActivity() {
+    private val MAX_CHAR_LEN = 16
     private val model: MainViewModel by viewModels()
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
-
+        enableStrictMode()
         val colorCircleView = findViewById<SelectedColorView>(R.id.colorCircle)
         val drawingFieldView = findViewById<DrawingFieldView>(R.id.drawField)
 
+        drawingFieldView.onFinishLine = {
+            model.activeBitmap = it
+        }
+        findViewById<TextView>(R.id.newLayer).setOnClickListener {
+            showLayerAlertDialog()
+        }
+        if (model.activeBitmap != null) {
+            model.activeBitmap?.let { it -> drawingFieldView.setBitmap(it) }
+            drawingFieldView.invalidate()
+        }
+
         colorCircleView.onColorChange = { stateColor ->
             model.primaryColor = stateColor
             drawingFieldView.updateColor(stateColor)
         }
 
+        if (model.dialogOpen) showLayerAlertDialog()
+        //model.viewModelScope
+
+    }
+
+    fun showLayerAlertDialog() {
+        val inflater = LayoutInflater.from(this@MainActivity)
+        val view = inflater.inflate(R.layout.layer_dialog, null)
+        val layerField = view.findViewById<EditText>(R.id.newLayerField)
+        val errorText = view.findViewById<TextView>(R.id.dialogCharCountError)
+
+        //if the layer number is not in the list already then set it to the field
+        if (model.layerNames.indexOf("Layer ${model.layerCounter}") == -1) {
+            layerField.setText(getString(R.string.new_layer_hint, model.layerCounter))
+        } else {
+            //else if it is in the list, keep looping till we find that its not
+            while (model.layerNames.indexOf("Layer ${model.layerCounter}") != -1) {
+                model.layerCounter += 1
+            }
+            layerField.setText(getString(R.string.new_layer_hint, model.layerCounter))
+        }
+        //needs to be after the if or else the textchanged listener will overwrite the previous text needed for rotations
+        layerField.addTextChangedListener {
+            model.dialogText = layerField.text.toString()
+            if (it.toString().trim().length > MAX_CHAR_LEN) {
+                errorText.visibility = View.VISIBLE
+                layerField.setText(layerField.text.toString().dropLast(1))
+            }
+        }
+
+        val dialog = AlertDialog.Builder(this).apply {
+            setView(view)
+        }.create()
+
+        view.findViewById<TextView>(R.id.newLayerCancel).setOnClickListener {
+            dialog.dismiss()
+        }
+        view.findViewById<TextView>(R.id.newLayerOk).setOnClickListener {
+            errorText.visibility = View.INVISIBLE
+            if (layerField.text.toString() ==
+                getString(R.string.new_layer_hint, model.layerCounter)
+            ) model.layerCounter += 1
```

Comment:
brackets around this

Reply:
I'm not sure what you mean, do you mean around the model.layerCounter?

Reply:
Yes. You have an if statement here, let's put brackets around its contents

Reply:
implemented